sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -656,3 +705,301 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
+        plan: LogicalPlan,
+        parentOuterReferences: AttributeSet,
+        aggregated: Boolean = false): ReturnType = {
+      val isCorrelated = hasOuterReferences(plan)
+      if (!isCorrelated) {
+        // We have reached a plan without correlation to the outer plan.
+        if (parentOuterReferences.isEmpty) {
+          // If there is no outer references from the parent nodes, it means all outer
+          // attributes can be substituted by attributes from the inner plan. So no
+          // domain join is needed.
+          (plan, Nil, Map.empty[Attribute, Attribute])
+        } else {
+          // Build the domain join with the parent outer references.
+          val attributes = parentOuterReferences.toSeq
+          val domains = attributes.map(_.newInstance())
+          // A placeholder to be rewritten into domain join.
+          val domainJoin = DomainJoin(domains, plan)
+          val outerReferenceMap = attributes.zip(domains).toMap
+          // Build join conditions between domain attributes and outer references.
+          // EqualNullSafe is used to make sure null key can be joined together. Note
+          // outer referenced attributes can be changed during the outer query optimization.
+          // The equality conditions will also serve as an attribute mapping between new
+          // outer references and domain attributes when rewriting the domain joins.
+          // E.g. if the attribute a is changed to a1, the join condition a' <=> outer(a)
+          // will become a' <=> a1, and we can construct the aliases based on the condition:
+          // DomainJoin [a']        Join Inner
+          // +- InnerQuery     =>   :- InnerQuery
+          //                        +- Aggregate [a1] [a1 AS a']
+          //                           +- OuterQuery
+          val conditions = outerReferenceMap.map {
+            case (o, a) => EqualNullSafe(a, OuterReference(o))
+          }
+          (domainJoin, conditions.toSeq, outerReferenceMap)
+        }
+      } else {
+        // Collect outer references from the current node.
+        val outerReferences = collectOuterReferences(plan.expressions)
+        plan match {
+          case Filter(condition, child) =>
+            val (correlated, uncorrelated) =
+              splitConjunctivePredicates(condition)
+              .partition(containsOuter)
+            val (equality, nonEquality) = correlated.partition(isEquality)
+            // Find equivalent outer reference relations and remove equivalent attributes from
+            // parentOuterReferences since they can be replaced directly by expressions
+            // inside the inner plan.
+            val equivalences = collectEquivalentOuterReferences(equality)
+            // When the results are aggregated, outer references inside the non-equality
+            // predicates cannot be used directly as join conditions with the outer query.
+            val outerReferences = if (aggregated) {
```

Comment:
let's avoid name conflicts with the outer `outerReferences`

Reply:
or we did it on purpose?

Reply:
Filter is the only node (for now) that doesn't use the outer `outerReferences`. I can move the outer computation into each case if that's clearer.